### ISSUE TITLE
XLNET completely wrong and random output

### ISSUE THREAD

**Issue by Oxi84** [huggingface/transformers#846]
I followed the example here: https://huggingface.co/pytorch-transformers/model_doc/xlnet.html#pytorch_transformers.XLNetModel

I found that I get completelly wrong output, I mean predicted word for the masked sentences are completelly irelevant and they change each run. I guess there is some bug, culd you please take a look at this:

**code:**
##############################
    
    config = XLNetConfig.from_pretrained('xlnet-large-cased')
    tokenizer = XLNetTokenizer.from_pretrained('xlnet-large-cased')'
    model = XLNetLMHeadModel(config)
    input_ids = torch.tensor(tokenizer.encode("Hello, my dog is very <..mask..> ")).unsqueeze(0)  # We will 
    predict the masked token
    print("input_ids",input_ids)
    perm_mask = torch.zeros((1, input_ids.shape[1], input_ids.shape[1]), dtype=torch.float)
    perm_mask[:, :, -1] = 1.0  # Previous tokens don't see last token
    target_mapping = torch.zeros((1, 1, input_ids.shape[1]), dtype=torch.float)  # Shape [1, 1, seq_length] => let's predict one token
    target_mapping[0, 0, -1] = 1.0 
    predictions = model(input_ids, perm_mask=perm_mask, target_mapping=target_mapping)

    predicted_k_indexes = torch.topk(predictions[0],k=10)
    predicted_logits_list = predicted_k_indexes[0] 
    predicted_indexes_list = predicted_k_indexes[1]    

    print ("predicted <masked> words:")
    for i,item  in enumerate(predicted_indexes_list[0][0]):
        the_index = predicted_indexes_list[0][0][i].item()
        print("word and logits",tokenizer.decode(the_index),predicted_logits_list[0][0][i].item())

###########################

output (one example - it changes each run):
#################################
input_ids tensor([[   17, 11368,    19,    94,  2288,    27,   172,     6]])
predicted <masked> words:

word and logits **emptiness** 2.7753820419311523
word and logits **Oklahoma** 2.61531400680542
word and logits **stars** 2.56619930267334
word and logits **bite** 2.5252184867858887
word and logits **Conte** 2.4745044708251953
word and logits **enforced** 2.4537196159362793
word and logits **antibody** 2.4416041374206543
word and logits **Got** 2.332545280456543
word and logits **Chev** 2.31380033493042
word and logits **MAG** 2.3047127723693848

####################################

**Comment by Oxi84** [comment]
It works much quite ok, when I use another way to load the model:
    
    tokenizer = XLNetTokenizer.from_pretrained("xlnet-large-cased")
    model = XLNetLMHeadModel.from_pretrained("xlnet-large-cased")
model.eval()

**Comment by Oxi84** [comment]
I also did comparasion with Bert, so far just one example, but I found  that Bert is much much better. I am not sure why is that.

**Comment by domaala** [comment]
Agreed! There is a chance we are not using the permutation mask and target mapping correctly, but I am suspicious as the documentation's example is not working very well either.

**Comment by thomwolf** [comment]
The main reason you get bad performance is that XLNet is not good on short inputs (comes from the way it is pretrained, always having a long memory and only guessing a few words in the sequence).

The `run_generation` example [here](https://github.com/huggingface/pytorch-transformers/blob/master/examples/run_generation.py) will show you how to get better performances by adding a random text as initiator.

Aman Rusia also wrote a blog post about that [here](https://medium.com/@amanrusia/xlnet-speaks-comparison-to-gpt-2-ea1a4e9ba39e). We are using his solution in the `run_generation` example.

**Comment by Oxi84** [comment]
Thanks, I am going to try the generation method and post the results here. Hope prediction is going to improve, but i guess that ading a lot of padding is make to slow the execution down a lot.

**Comment by cpcdoy** [comment]
@Oxi84 Any luck with your results ? I still get pretty random results even while using this trick

**Comment by seungbochoi** [comment]
hey guys, can someone give some more insight what @thomwolf talking about? I can't understand the difference the way both Bert and XLnetLM works for LMhead task. Aren't both model having disadvantages if they have short sentence? It seems he said XLnet has disadvantage on short input sentence while Bert does not(or has less disadvantage). Any detail explanation could be useful ! 

Also, I can't get the point of adding extra padding or adding random padding things to improve XLnetLMHead model. Any snippet or explanation could be appreciated too...(saw the link but could not fully understood)

**Comment by iedmrc** [comment]
I think these questions are not directly related to this repo. Maybe you should check out the [paper](https://arxiv.org/abs/1906.08237) or ask on [quora](https://www.quora.com/) or on [researchgate](https://explore.researchgate.net/display/support/Asking+questions)
